File: TB_IPR/TUT.IMG.point_processes_generation/matlab/ripley.m
function [K, L, vals]=ripley(x,y,xmin,xmax,ymin,ymax,edges)
% Ripley K function
% x and y define the points abscisses and ordinates
% xmin xmax ymin ymax define the domain
% edges is a vector containing the distances (typically, r=1:100 for example)
% K: K function
% L: L function
% vals: values of radius

% number of points
nb_points=length(x);

% area
area = (xmax-xmin)*(ymax-ymin);

% computes the distances between all pairs of points
d = pdist([x,y]);

h = histcounts(d, edges, 'Normalization', 'cumcount');
figure, plot(h);

% count mean number of points
% multiply by 2 because each distance is only counted once in g
K = 2*h/nb_points;

% estimates density
densite=nb_points/area;
K=K/densite;

L = sqrt(K/pi);

vals = edges(1:end-1)+diff(edges);
%figure, plot(vals, h); title('cumulative histogram')
end % of function

File: TB_IPR/TUT.IMG.point_processes_generation/python/ripley.py
# -*- coding: utf-8 -*-
"""
Created on Wed Nov  2 15:30:14 2016

@author: yann
"""
import numpy as np
from scipy.spatial.distance import pdist
import matplotlib.pyplot as plt


def ripley(x, y, xmin, xmax, ymin, ymax, edges):
    """
    Ripley K and L functions, vals is values of radius
    this function has border effects!
    x, y: coordinates of points
    xmin, xmax, ymin, ymax: window
    edges: values of bins for histogram evaluation
    """

    # number of points
    nb_points = x.size

    # compute pairwise distances
    P = np.transpose(np.vstack((x, y)))
    d = pdist(P)

    # compute cumulative histogram
    h, edges = np.histogram(d, edges)
    H = np.cumsum(h)

    # normalization of K
    K = 2*H/nb_points
    area = (xmax-xmin) * (ymax-ymin)
    density = float(nb_points) / area
    K = K / density

    # L
    L = np.sqrt(K/np.pi)

    # edge values
    vals = edges[:-1] + np.diff(edges)

    return K, L, vals
